agora/Agora/Proposal/Scripts.hs

```diff
@@ -600,81 +600,82 @@ proposalValidator as maximumCosigners =
 
           ----------------------------------------------------------------------
 
-          PAdvanceProposal _ ->
-            let currentTime' = pfromJust # currentTime
-                fromDraft = withMultipleStakes $ \totalStakedAmount sortedStakeOwners ->
-                  pmatchC (isDraftPeriod # proposalF.timingConfig # proposalF.startingTime # currentTime') >>= \case
-                    PTrue -> do
-                      pguardC "More cosigns than minimum amount" $
-                        punsafeCoerce (pfromData thresholdsF.vote) #< totalStakedAmount
-
-                      pguardC "All new cosigners are witnessed by their Stake datums" $
-                        plistEquals # sortedStakeOwners # proposalF.cosigners
-
-                      -- 'Draft' -> 'VotingReady'
-                      pguardC "Proposal status set to VotingReady" $
-                        proposalOutStatus #== pconstant VotingReady
-
-                      pure $ pconstant ()
-                    PFalse -> do
-                      pguardC "Advance to failed state" $ proposalOutStatus #== pconstant Finished
-
-                      pure $ pconstant ()
-
-                fromOther = withSingleStake $ \_ _ stakeUnchanged -> do
-                  pguardC "Stake should not change" stakeUnchanged
+          PAdvanceProposal _ -> unTermCont $ do
+            currentTime' <- pletC $ pfromJust # currentTime
+            let inDraftPeriod = isDraftPeriod # proposalF.timingConfig # proposalF.startingTime # currentTime'
+            inVotingPeriod <- pletC $ isVotingPeriod # proposalF.timingConfig # proposalF.startingTime # currentTime'
+            inLockedPeriod <- pletC $ isLockingPeriod # proposalF.timingConfig # proposalF.startingTime # currentTime'
+            inExecutionPeriod <- pletC $ isExecutionPeriod # proposalF.timingConfig # proposalF.startingTime # currentTime'
+            pguardC "Only status changes in the output proposal" onlyStatusChanged
+            let gstSymbol = pconstant $ governorSTSymbol as
+            gstMoved <-
+              pletC $
+                pany
+                  # plam
+                    ( \( (pfield @"value" #)
+                          . (pfield @"resolved" #) ->
+                          value
+                        ) ->
+                          psymbolValueOf # gstSymbol # value #== 1
+                    )
+                  # pfromData txInfoF.inputs
+            let toFailedState = unTermCont $ do
+                  pguardC "Proposal should fail: not on time" $
+                    proposalOutStatus #== pconstant Finished
+
+                  pguardC "GST not moved" $ pnot # gstMoved
+
+                  pure $ pconstant ()
+            pure $
+              pmatchEnum (pto currentStatus) $ \case
+                Draft ->
```

Comment:
...And use plutarch level constructor here.

Reply:
Right! Should be equally efficient.